From ecb292cc3719695d2b27617ddee9cab1895d3e17 Mon Sep 17 00:00:00 2001
From: Dave Eddy <dave@daveeddy.com>
Date: Tue, 3 Oct 2017 12:23:07 -0400
Subject: [PATCH] OS-6300 vmadm and vmadmd create zonecfg race Reviewed by:
 Patrick Mooney <patrick.mooney@joyent.com> Reviewed by: Josh Wilsdon
 <josh@wilsdon.ca> Approved by: Josh Wilsdon <josh@wilsdon.ca>

---
 src/Makefile                   |   1 +
 src/manifest                   |   1 +
 src/vm/node_modules/VM.js      | 202 ++++++++++++++-------------------
 src/vm/node_modules/zonecfg.js | 163 ++++++++++++++++++++++++++
 src/vm/sbin/vmadmd.js          | 118 ++++++++++---------
 5 files changed, 312 insertions(+), 173 deletions(-)
 create mode 100644 src/vm/node_modules/zonecfg.js

diff --git a/src/Makefile b/src/Makefile
index 8d217572..d818111a 100644
--- a/src/Makefile
+++ b/src/Makefile
@@ -93,6 +93,7 @@ JS_CHECK_TARGETS=\
 	vm/node_modules/openonerrlogger.js \
 	vm/node_modules/vmload/*.js \
 	vm/node_modules/sysevent-stream.js \
+	vm/node_modules/zonecfg.js \
 	img/lib/*.js \
 	img/sbin/imgadm \
 	vm/common/nictag.js \
diff --git a/src/manifest b/src/manifest
index b60f314c..a4cbede8 100644
--- a/src/manifest
+++ b/src/manifest
@@ -545,6 +545,7 @@ f usr/vm/node_modules/vmload/vmload-xml.js 0644 root root
 f usr/vm/node_modules/vmload/dump-json.js 0755 root root
 f usr/vm/node_modules/vmload/index.js 0644 root root
 f usr/vm/node_modules/VM.js 0444 root bin
+f usr/vm/node_modules/zonecfg.js 0444 root bin
 f usr/vm/smf/system-vmadmd 0555 root bin
 f usr/vm/smf/system-vmadmd.xml 0444 root bin
 # fwadm
diff --git a/src/vm/node_modules/VM.js b/src/vm/node_modules/VM.js
index 92b4e1f5..82253d96 100644
--- a/src/vm/node_modules/VM.js
+++ b/src/vm/node_modules/VM.js
@@ -98,6 +98,7 @@ var util = require('util');
 var utils = require('./utils');
 var vasync = require('/usr/vm/node_modules/vasync');
 var vmload = require('vmload');
+var zonecfg = require('/usr/vm/node_modules/zonecfg');
 
 // pull in stuff from generated props (originating in proptable.js)
 var BRAND_OPTIONS = properties.BRAND_OPTIONS;
@@ -2128,6 +2129,7 @@ function createHostConfFileMounts(vmobj, opts, log, callback) {
     var tracers_obj;
 
     assert.object(vmobj, 'vmobj');
+    assert.uuid(vmobj.uuid, 'vmobj.uuid');
     assert.object(opts, 'opts');
     assert.object(log, 'log');
     assert.func(callback, 'callback');
@@ -2351,7 +2353,7 @@ function createHostConfFileMounts(vmobj, opts, log, callback) {
 
         zcfg = buildFilesystemZonecfg({}, fake_payload);
 
-        zonecfgFile(zcfg, ['-z', vmobj.zonename], log,
+        zonecfg(vmobj.uuid, [], {log: log, stdin: zcfg},
             function (zcfg_err, fds) {
                 if (zcfg_err) {
                     log.error({
@@ -2565,26 +2567,27 @@ function createFilesystems(payload, filesystems, log, callback)
                 cb(new Error('createFilesystems() do not recognize source'));
             }
         }, function (err) {
-            // send the zonecfg data we just generated as a file to zonecfg,
+            // send the zonecfg data we just generated as stdin to zonecfg,
             // this will create the zone.
-            zonecfgFile(zcfg, ['-z', payload.zonename], log,
+            zonecfg(payload.zonename, [],
+                {log: log, stdin: zcfg, useZonename: true},
                 function (zcfg_err, fds) {
-                    if (zcfg_err) {
-                        log.error({
-                            err: zcfg_err,
-                            zcfg: zcfg,
-                            stdout: fds.stdout,
-                            stderr: fds.stderr
-                        }, 'failed to modify zonecfg');
-                        callback(zcfg_err);
-                        return;
-                    }
 
-                    log.debug({stdout: fds.stdout, stderr: fds.stderr},
-                        'modified zonecfg');
-                    callback();
+                if (zcfg_err) {
+                    log.error({
+                        err: zcfg_err,
+                        zcfg: zcfg,
+                        stdout: fds.stdout,
+                        stderr: fds.stderr
+                    }, 'failed to modify zonecfg');
+                    callback(zcfg_err);
+                    return;
                 }
-            );
+
+                log.debug({stdout: fds.stdout, stderr: fds.stderr},
+                    'modified zonecfg');
+                callback();
+            });
         });
     });
 }
@@ -3392,58 +3395,6 @@ function writeZoneconfig(payload, log, callback)
     }
 }
 
-function zonecfg(args, log, callback)
-{
-    var cmd = '/usr/sbin/zonecfg';
-
-    assert(log, 'no logger passed to zonecfg()');
-
-    traceExecFile(cmd, args, log, 'zonecfg', function (error, stdout, stderr) {
-        if (error) {
-            callback(error, {'stdout': stdout, 'stderr': stderr});
-        } else {
-            callback(null, {'stdout': stdout, 'stderr': stderr});
-        }
-    });
-}
-
-function zonecfgFile(data, args, log, callback)
-{
-    var tmpfile = '/tmp/zonecfg.' + process.pid + '.tmp';
-    var tracers_obj;
-
-    assert(log, 'no logger passed to zonecfgFile()');
-
-    if (process.env.EXPERIMENTAL_VMJS_TRACING) {
-        tracers_obj = traceUntilCallback('zonecfg-file', log, callback);
-        callback = tracers_obj.callback;
-        log = tracers_obj.log;
-    }
-
-    log.debug({data: data}, tmpfile + ' contents');
-
-    fs.writeFile(tmpfile, data, function (err, result) {
-        if (err) {
-            // On failure we don't delete the tmpfile so we can debug it.
-            callback(err);
-        } else {
-            args.push('-f');
-            args.push(tmpfile);
-
-            zonecfg(args, log, function (e, fds) {
-                if (e) {
-                    // keep temp file around for investigation
-                    callback(e, fds);
-                } else {
-                    fs.unlink(tmpfile, function () {
-                        callback(null, fds);
-                    });
-                }
-            });
-        }
-    });
-}
-
 function zoneadm(args, log, callback)
 {
     var cmd = '/usr/sbin/zoneadm';
@@ -5131,7 +5082,9 @@ function createDelegatedDataset(payload, log, callback)
             }
 
             zcfg = zcfg + 'add dataset; set name=' + ds + '; end\n';
-            zonecfg(['-u', payload.uuid, zcfg], log, function (e, fds) {
+            zonecfg(payload.uuid, [zcfg], {log: log},
+                function (e, fds) {
+
                 if (e) {
                     log.error({'err': e, stdout: fds.stdout,
                         stderr: fds.stderr}, 'unable to add delegated dataset '
@@ -6715,7 +6668,8 @@ exports.markVMFailure = function (vmobj, options, callback)
         zcfg = 'remove -F attr name=failed; add attr; set name=failed; '
             + 'set value="provisioning"; set type=string; end';
 
-        zonecfg(['-u', vmobj.uuid, zcfg], log, function (zonecfg_err, fds) {
+        zonecfg(vmobj.uuid, [zcfg], {log: log},
+            function (zonecfg_err, fds) {
 
             if (zonecfg_err) {
                 log.error({err: zonecfg_err, stdout: fds.stdout,
@@ -8155,9 +8109,11 @@ function createZone(payload, log, callback)
     // we're talking about a new machine.
     zcfg = zcfg + buildZonecfgUpdate({}, payload, log);
 
-    // send the zonecfg data we just generated as a file to zonecfg,
+    // send the zonecfg data we just generated as stdin to zonecfg,
     // this will create the zone.
-    zonecfgFile(zcfg, ['-z', payload.zonename], log, function (err, fds) {
+    zonecfg(payload.zonename, [],
+        {log: log, stdin: zcfg, useZonename: true}, function (err, fds) {
+
         if (err) {
             log.error({err: err, zcfg: zcfg, stdout: fds.stdout,
                 stderr: fds.stderr}, 'failed to modify zonecfg');
@@ -8648,8 +8604,8 @@ exports.unsetTransition = function (vmobj, options, callback)
         log = tracers_obj.log;
     }
 
-    zonecfg(['-u', vmobj.uuid, 'remove -F attr name=transition'], log,
-        function (err, fds) {
+    zonecfg(vmobj.uuid, ['remove -F attr name=transition'],
+        {log: log}, function (err, fds) {
 
         if (err) {
             // log at info because this might be because already removed
@@ -8660,8 +8616,8 @@ exports.unsetTransition = function (vmobj, options, callback)
                 'removed transition for zone ' + vmobj.uuid);
         }
 
-        zonecfg(['-u', vmobj.uuid, 'info attr name=transition'], log,
-            function (info_err, info_fds) {
+        zonecfg(vmobj.uuid, ['info attr name=transition'],
+            {log: log}, function (info_err, info_fds) {
 
             if (info_err) {
                 log.error({err: info_err, stdout: info_fds.stdout,
@@ -8732,7 +8688,9 @@ function setTransition(vmobj, transition, target, timeout, log, callback)
             var zcfg;
 
             zcfg = buildTransitionZonecfg(transition, target, timeout);
-            zonecfg(['-u', vmobj.uuid, zcfg], log, function (err, fds) {
+            zonecfg(vmobj.uuid, [zcfg], {log: log},
+                function (err, fds) {
+
                 if (err) {
                     log.error({err: err, stdout: fds.stdout,
                         stderr: fds.stderr}, 'failed to set transition='
@@ -9469,7 +9427,7 @@ exports.reprovision = function (uuid, payload, options, callback)
             // update zone's image_uuid field
             var zcfg = 'select attr name=dataset-uuid; set value="'
                 + payload.image_uuid + '"; end';
-            zonecfg(['-u', uuid, zcfg], log, function (err, fds) {
+            zonecfg(uuid, [zcfg], {log: log}, function (err, fds) {
                 if (err) {
                     log.error({err: err, stdout: fds.stdout,
                         stderr: fds.stderr}, 'unable to set image_uuid on VM '
@@ -10203,7 +10161,9 @@ function deleteZone(uuid, log, callback)
         // TODO: replace these next two with VM.stop(..{force: true} ?
         }, function (cb) {
             log.debug('setting autoboot=false');
-            zonecfg(['-u', uuid, 'set autoboot=false'], log, function (e, fds) {
+            zonecfg(uuid, ['set autoboot=false'], {log: log},
+                function (e, fds) {
+
                 if (e) {
                     log.warn({err: e, stdout: fds.stdout, stderr: fds.stderr},
                         'Error setting autoboot=false');
@@ -10271,24 +10231,24 @@ function deleteZone(uuid, log, callback)
                 cb();
             }
         }, function (cb) {
-            if (vmobj.zonename) {
-                log.debug('deleting zone');
-                // XXX for some reason -u <uuid> doesn't work with delete
-                zonecfg(['-z', vmobj.zonename, 'delete', '-F'], log,
-                    function (e, fds) {
-
-                    if (e) {
-                        log.warn({err: e, stdout: fds.stdout,
-                            stderr: fds.stderr}, 'Error deleting VM');
-                    } else {
-                        log.debug({stdout: fds.stdout, stderr: fds.stderr},
-                            'deleted VM ' + uuid);
-                    }
-                    cb();
-                });
-            } else {
+            if (!vmobj.zonename) {
                 cb();
+                return;
             }
+
+            log.debug('deleting zone');
+            zonecfg(uuid, ['delete', '-F'],
+                {log: log}, function (e, fds) {
+
+                if (e) {
+                    log.warn({err: e, stdout: fds.stdout,
+                        stderr: fds.stderr}, 'Error deleting VM');
+                } else {
+                    log.debug({stdout: fds.stdout, stderr: fds.stderr},
+                        'deleted VM ' + uuid);
+                }
+                cb();
+            });
         }, function (cb) {
             VM.load(uuid, {fields: ['uuid'], log: log, missing_ok: true},
                 function (err, obj) {
@@ -10568,7 +10528,7 @@ function startZone(vmobj, opts, callback)
                 }, 'docker VM has restart policy, setting autoboot');
             }
 
-            zonecfg(['-u', uuid, set_autoboot], log,
+            zonecfg(uuid, [set_autoboot], {log: log},
                 function (err, autoboot_fds) {
 
                 if (err) {
@@ -10588,21 +10548,21 @@ function startZone(vmobj, opts, callback)
                 cb();
                 return;
             }
-            zonecfg(['-u', uuid, 'remove attr name=never-booted' ], log,
-                function (err, neverbooted_fds) {
-                    // Ignore errors here, because we're started.
-                    if (err) {
-                        log.warn({err: err, stdout: neverbooted_fds.stdout,
-                            stderr: neverbooted_fds.stderr}, 'failed to remove '
-                            + 'never-booted flag');
-                    } else {
-                        log.debug({stdout: neverbooted_fds.stdout,
-                            stderr: neverbooted_fds.stderr}, 'removed '
-                            + 'never-booted flag');
-                    }
-                    cb();
+            zonecfg(uuid, ['remove attr name=never-booted'],
+                {log: log}, function (err, neverbooted_fds) {
+
+                // Ignore errors here, because we're started.
+                if (err) {
+                    log.warn({err: err, stdout: neverbooted_fds.stdout,
+                        stderr: neverbooted_fds.stderr}, 'failed to remove '
+                        + 'never-booted flag');
+                } else {
+                    log.debug({stdout: neverbooted_fds.stdout,
+                        stderr: neverbooted_fds.stderr}, 'removed '
+                        + 'never-booted flag');
                 }
-            );
+                cb();
+            });
         }
     ], function (err) {
         if (!err) {
@@ -13452,10 +13412,12 @@ exports.update = function (uuid, payload, options, callback)
             });
         }, function (cb) {
             var zcfg;
-            // generate a payload and send as a file to zonecfg to update
-            // the zone.
+            // generate a payload and send as stdin to zonecfg to update the
+            // zone.
             zcfg = buildZonecfgUpdate(vmobj, payload, log);
-            zonecfgFile(zcfg, ['-u', uuid], log, function (e, fds) {
+            zonecfg(uuid, [], {log: log, stdin: zcfg},
+                function (e, fds) {
+
                 if (e) {
                     log.error({err: e, stdout: fds.stdout, stderr: fds.stderr},
                         'unable to update zonecfg');
@@ -13568,7 +13530,7 @@ function halt(uuid, log, callback)
             } else {
                 log.debug({stdout: fds.stdout, stderr: fds.stderr},
                     'zoneadm halted VM ' + uuid);
-                zonecfg(['-u', uuid, unset_autoboot], log,
+                zonecfg(uuid, [unset_autoboot], {log: log},
                     function (error, unset_fds) {
 
                     if (error) {
@@ -13840,7 +13802,7 @@ function doDockerStop(vmobj, options, callback)
         });
     }
 
-    zonecfg(['-u', vmobj.uuid, unset_autoboot], log,
+    zonecfg(vmobj.uuid, [unset_autoboot], {log: log},
         function (zonecfg_err, fds) {
 
         if (zonecfg_err) {
@@ -13963,7 +13925,7 @@ function doVmadmdStop(vmobj, options, callback)
             var unset_autoboot = 'select attr name=vm-autoboot; '
                 + 'set value=false; end';
 
-            zonecfg(['-u', vmobj.uuid, unset_autoboot], log,
+            zonecfg(vmobj.uuid, [unset_autoboot], {log: log},
                 function (err, fds) {
                     if (err) {
                         // The vm is dead at this point, failing
@@ -14026,7 +13988,7 @@ function doShutdownStop(vmobj, options, callback)
                 cb();
             });
         }, function (cb) {
-            zonecfg(['-u', vmobj.uuid, unset_autoboot], log,
+            zonecfg(vmobj.uuid, [unset_autoboot], {log: log},
                 function (err, fds) {
 
                 if (err) {
@@ -14140,7 +14102,9 @@ exports.stop = function (uuid, options, callback)
             }
 
             log.debug('setting autoboot=false');
-            zonecfg(['-u', uuid, unset_autoboot], log, function (e, fds) {
+            zonecfg(uuid, [unset_autoboot], {log: log},
+                function (e, fds) {
+
                 if (e) {
                     log.warn({err: e, stdout: fds.stdout, stderr: fds.stderr},
                         'Error setting autoboot=false');
diff --git a/src/vm/node_modules/zonecfg.js b/src/vm/node_modules/zonecfg.js
new file mode 100644
index 00000000..e7ab23eb
--- /dev/null
+++ b/src/vm/node_modules/zonecfg.js
@@ -0,0 +1,163 @@
+/*
+ * CDDL HEADER START
+ *
+ * The contents of this file are subject to the terms of the
+ * Common Development and Distribution License, Version 1.0 only
+ * (the "License").  You may not use this file except in compliance
+ * with the License.
+ *
+ * You can obtain a copy of the license at http://smartos.org/CDDL
+ *
+ * See the License for the specific language governing permissions
+ * and limitations under the License.
+ *
+ * When distributing Covered Code, include this CDDL HEADER in each
+ * file.
+ *
+ * If applicable, add the following below this CDDL HEADER, with the
+ * fields enclosed by brackets "[]" replaced with your own identifying
+ * information: Portions Copyright [yyyy] [name of copyright owner]
+ *
+ * CDDL HEADER END
+ *
+ * Copyright 2017, Joyent, Inc.
+ *
+ * Zonecfg wrapper with a per-zone lock file
+ */
+
+var cp = require('child_process');
+var util = require('util');
+
+var assert = require('/usr/node/node_modules/assert-plus');
+var lock = require('/usr/vm/node_modules/locker').lock;
+var vasync = require('/usr/vm/node_modules/vasync');
+
+var LOCK_PATH = '/var/run/.vm.%s.zonecfg.lockfile';
+
+module.exports = zonecfg;
+
+/*
+ * Call zonecfg(1M) with a given set of arguments and optional stdin data.
+ * A lock file will be acquired (based on the id given as the first argument,
+ * typically a zoneuuid) to ensure that only one instance of zonecfg will be
+ * run at a time per zone.
+ *
+ * Example:
+ *
+ * var uuid = 'ac1946a2-fa79-47c7-9bc3-729fcc5a19d2';
+ * var cmd = ['set autoboot=false'];
+ * var opts = {
+ *   log: log
+ * };
+ *
+ * zonecfg(uuid, cmd, opts, function (err, fds) {
+ *     if (err) {
+ *         // command exited non-zero or failed to spawn for some reason
+ *         throw err;
+ *     }
+ *
+ *     // fds.stdout and fds.stderr set
+ * });
+ *
+ * - uuid: a zone's UUID (preferable) or zonename
+ * - args: arguments to pass to zonecfg(1M) as an array of strings
+ * - opts: options object
+ *   - opts.log         [required] bunyan logger
+ *   - opts.stdin       [optional] a string to pass to the child as stdin
+ *   - opts.useZonename [optional] a boolean that tells if the uuid argument is
+ *                                 the zones uuid or zonename
+ * - callback: callback function with the signature function (err, fds) {}
+ *
+ * By default, the arguments ['-u', uuid] will be prepended to the arguments
+ * passed with the second argument.  If opts.useZonename is true, then ['-z',
+ * uuid] will be specified instead - this is particularly useful when creating
+ * a zone, since the zone will not exist and thus not have a UUID to be
+ * referenced.
+ *
+ */
+function zonecfg(uuid, args, opts, callback) {
+    assert.uuid(uuid, 'uuid');
+    assert.arrayOfString(args, 'args');
+    assert.object(opts, 'opts');
+    assert.object(opts.log, 'opts.log');
+    assert.optionalString(opts.stdin, 'opts.stdin');
+    assert.optionalBool(opts.useZonename, 'opts.useZonename');
+    assert.func(callback, 'callback');
+
+    args.unshift(uuid);
+    args.unshift(opts.useZonename ? '-z' : '-u');
+
+    var cmd = 'zonecfg';
+    var file = util.format(LOCK_PATH, uuid);
+    var child;
+    var unlock;
+    var execOpts = {
+        encoding: 'utf8'
+    };
+    var ret = {};
+    var log = opts.log.child({cmd: cmd, args: args, file: file});
+
+    vasync.pipeline({funcs: [
+        function (_, cb) {
+            // acquire lock
+            log.debug('acquiring lock %s', file);
+
+            lock(file, function (err, _unlock) {
+                if (err) {
+                    log.error({err: err}, 'failed to acquire lock %s', file);
+                    cb(err);
+                    return;
+                }
+
+                log.debug('acquired lock %s', file);
+
+                unlock = _unlock;
+                cb();
+            });
+        }, function (_, cb) {
+            log.debug('calling %s', cmd);
+
+            child = cp.execFile(cmd, args, execOpts,
+                function (err, stdout, stderr) {
+
+                ret.stderr = stderr;
+                ret.stdout = stdout;
+
+                if (err) {
+                    log.error({err: err}, '%s call failed', cmd);
+                    cb(err);
+                    return;
+                }
+
+                log.debug({ret: ret}, '%s ran successfully', cmd);
+                cb();
+            });
+
+            if (opts.stdin !== undefined) {
+                log.debug({stdin: opts.stdin}, 'writing to stdin');
+                child.stdin.write(opts.stdin);
+                child.stdin.end();
+            }
+        }
+    ]}, function (err) {
+        if (err) {
+            log.error({err: err}, 'zonecfg error for %s: %s',
+                uuid, err.message);
+        }
+
+        if (!unlock) {
+            callback(err, ret);
+            return;
+        }
+
+        unlock(function (unlock_err) {
+            if (unlock_err) {
+                log.error({err: unlock_err}, 'failed to unlock %s', file);
+                callback(unlock_err);
+                return;
+            }
+
+            callback(err, ret);
+        });
+    });
+}
diff --git a/src/vm/sbin/vmadmd.js b/src/vm/sbin/vmadmd.js
index 734445f2..0cd5518d 100755
--- a/src/vm/sbin/vmadmd.js
+++ b/src/vm/sbin/vmadmd.js
@@ -29,11 +29,9 @@ var assert = require('/usr/node/node_modules/assert-plus');
 var async = require('/usr/node/node_modules/async');
 var bunyan = require('/usr/node/node_modules/bunyan');
 var cp = require('child_process');
-var consts = require('constants');
 var EventEmitter = require('events').EventEmitter;
 var execFile = cp.execFile;
 var fs = require('fs');
-var lock = require('/usr/vm/node_modules/locker').lock;
 var mod_nic = require('/usr/vm/node_modules/nic');
 var net = require('net');
 var VM = require('/usr/vm/node_modules/VM');
@@ -46,6 +44,7 @@ var SyseventStream = require('/usr/vm/node_modules/sysevent-stream');
 var url = require('url');
 var util = require('util');
 var vasync = require('vasync');
+var zonecfg = require('/usr/vm/node_modules/zonecfg');
 
 /*
  * The DOCKER_RUNTIME_DELAY_RESET parameter is used when restarting a Docker VM
@@ -119,20 +118,6 @@ function zfs(args, callback)
     });
 }
 
-function zonecfg(args, callback)
-{
-    var cmd = '/usr/sbin/zonecfg';
-
-    log.debug(cmd + ' ' + args.join(' '));
-    execFile(cmd, args, function (error, stdout, stderr) {
-        if (error) {
-            callback(error, {'stdout': stdout, 'stderr': stderr});
-        } else {
-            callback(null, {'stdout': stdout, 'stderr': stderr});
-        }
-    });
-}
-
 /*
  * This function gets the runtime of the last run of a stopped zone by comparing
  * the mtime of the /lastbooted and /lastexited files.
@@ -2057,6 +2042,7 @@ function upgradeVM(vmobj, fields, callback)
             });
         }, function (cb) {
             var args = [];
+            var cmd;
 
             if (vmobj.image_uuid || vmobj.brand === 'kvm') {
                 // already have image_uuid, no problem
@@ -2091,9 +2077,16 @@ function upgradeVM(vmobj, fields, callback)
 
                 image_uuid = origin.split('@')[0].split('/').pop();
                 log.info('setting new image_uuid: ' + image_uuid);
-                zonecfg(['-z', vmobj.zonename, 'add attr; '
-                    + 'set name=dataset-uuid; set type=string; set value="'
-                    + image_uuid + '"; end'],
+                cmd = [
+                    'add attr',
+                    'set name=dataset-uuid',
+                    'set type=string',
+                    'set value="' + image_uuid + '"',
+                    'end'
+                ].join('; ');
+
+                zonecfg(vmobj.uuid, [cmd], {log: log},
+
                     function (add_err, add_fds) {
                         if (add_err) {
                             log.error(add_err);
@@ -2199,22 +2192,22 @@ function upgradeVM(vmobj, fields, callback)
                 cb();
             });
         }, function (cb) {
-            if (vmobj.hasOwnProperty('default_gateway')) {
-                zonecfg(['-z', vmobj.zonename,
-                    'remove attr name=default-gateway'],
-                    function (err, fds) {
-                        if (err) {
-                            log.error(err);
-                            cb(err);
-                            return;
-                        }
-                        log.info('removed default-gateway');
-                        cb();
-                    }
-                );
-            } else {
+            if (!vmobj.hasOwnProperty('default_gateway')) {
                 cb();
+                return;
             }
+
+            zonecfg(vmobj.uuid, ['remove attr name=default-gateway'],
+                {log: log}, function (err, fds) {
+
+                if (err) {
+                    log.error(err);
+                    cb(err);
+                    return;
+                }
+                log.info('removed default-gateway');
+                cb();
+            });
         }, function (cb) {
             // for KVM we always want 10G zoneroot quota
             if (vmobj.brand === 'kvm' && vmobj.quota !== 10) {
@@ -2223,6 +2216,8 @@ function upgradeVM(vmobj, fields, callback)
             }
             cb();
         }, function (cb) {
+            var cmd;
+
             if (vmobj.hasOwnProperty('create_timestamp')) {
                 cb();
                 return;
@@ -2250,19 +2245,24 @@ function upgradeVM(vmobj, fields, callback)
 
                 log.info('creation time: ' + creation_timestamp + ' from ZFS');
 
-                zonecfg(['-z', vmobj.zonename, 'add attr; '
-                    + 'set name=create-timestamp; set type=string; '
-                    + 'set value="' + creation_timestamp + '"; end'],
+                cmd = [
+                    'add attr',
+                    'set name=create-timestamp',
+                    'set type=string',
+                    'set value="' + creation_timestamp + '"',
+                    'end'
+                ].join('; ');
+                zonecfg(vmobj.uuid, [cmd], {log: log},
                     function (zcfg_err, zcfg_fds) {
-                        if (zcfg_err) {
-                            log.error(zcfg_err);
-                            cb(zcfg_err);
-                            return;
-                        }
-                        log.info('set create-timestamp: ' + creation_timestamp);
-                        cb();
+
+                    if (zcfg_err) {
+                        log.error(zcfg_err);
+                        cb(zcfg_err);
+                        return;
                     }
-                );
+                    log.info('set create-timestamp: ' + creation_timestamp);
+                    cb();
+                });
             });
         }, function (cb) {
             // in SDC7 *_pw keys do not work in customer_metadata and must be in
@@ -2335,19 +2335,29 @@ function upgradeVM(vmobj, fields, callback)
             });
         }, function (cb) {
             // zonecfg update vm-version = 1
+            var cmd;
+
             log.debug('setting vm-version = 1');
-            zonecfg(['-z', vmobj.zonename, 'add attr; set name=vm-version; '
-                + 'set type=string; set value=1; end'],
+
+            cmd = [
+                'add attr',
+                'set name=vm-version',
+                'set type=string',
+                'set value=1',
+                'end'
+            ].join('; ');
+
+            zonecfg(vmobj.uuid, [cmd], {log: log},
                 function (err, fds) {
-                    if (err) {
-                        log.error(err);
-                        cb(err);
-                        return;
-                    }
-                    log.info('set vm-version = 1');
-                    cb();
+
+                if (err) {
+                    log.error(err);
+                    cb(err);
+                    return;
                 }
-            );
+                log.info('set vm-version = 1');
+                cb();
+            });
         }, function (cb) {
             // reload VM so we get all the updated properties
             VM.load(vmobj.uuid, {fields: fields, log: log},
-- 
2.21.0

